crates/bootstrap_srv/src/store.rs

```diff
@@ -0,0 +1,193 @@
+//! This is a virtual-memory inspired tempfile store.
+//!
+//! ### Rationale
+//!
+//! - We don't need to persist anything in the store beyond a single
+//!   process invocation, because the infos are going to expire after
+//!   a matter of minutes anyways, and peers will continue to re-publish them.
+//! - We would like a server to be able to store more spaces and infos
+//!   than would reasonably fit in RAM.
+//! - We would like to avoid use of an external database, to reduce
+//!   dependencies and ease deployment and testing.
+//!
+//! ### Implementation
+//!
+//! - Have a pool of tempfiles that can grow to match the worker thread
+//!   count if needed.
+//! - Have the ability to Clone (reopen) these tempfile handles.
+//! - Designate one file handle only per pool entry as writable.
+//! - All other handles will be readonly.
+//! - After writing an entry, return a reference with a readonly file handle
+//!   to the tempfile that was written to, and an offset/length to
+//!   allow readback of exactly the bytes written for that entry.
+//! - Once a tempfile reaches a certain size, drop the write handle and
+//!   open a new tempfile for writing.
+//! - The older read handles will persist the existence of the older tempfiles
+//!   until the last read reference is dropped, at which point the tempfile
+//!   will be cleaned up by the os.
+//! - We can trust these files will cycle through at a rate similar to the
+//!   max expiration time on the infos they contain (30 minutes).
```

Comment:
Almost. We're writing the individual agent infos to the store, and keeping the space list with only the references in memory. So when we update, only new infos are written to the disk. This avoids us re-writing the same infos over and over every time the space updates.

Yes indeed, the max expires_at is 30 min past the created_at and the max in the future created_at can be is 3 min, so the max we'll hold an individual info is 33 min plus the interval the prune task runs on.

Reply:
I see, so I'd slightly misunderstood what's being written. So that's more efficient than I was thinking :)

I see the logic of not holding the info for more than 33 minutes ish. However, that doesn't prevent rapid agent updates right? So I could send a new agent info every 5ms and it would force the space required by temporary files to grow rapidly. It would expire after 33 minutes so the files could be dropped but still, you could add quite a load before the first ones start expiring

Reply:
Your logic is correct. I'd say this will be mitigated by rate limiting.

We could potentially use a different strategy such as having a tempfile per space, but that inevitably increases system resource usage (like max file handles).

Reply:
Cool, yes that sounds like a rate limiting issue - happy for this to be solved elsewhere :)